packages/astro/src/core/routing/rewrite.ts

```diff
@@ -12,44 +12,65 @@ export type FindRouteToRewrite = {
 	base: AstroConfig['base'];
 };
 
+export interface FindRouteToRewriteResult {
+	routeData: RouteData;
+	newUrl: URL;
+	pathname: string;
+}
+
+/**
+ * Shared logic to retrieve the rewritten route. It returns a tuple that represents:
+ * 1. The new `Request` object. It contains `base`
+ * 2.
+ */
 export function findRouteToRewrite({
 	payload,
 	routes,
 	request,
 	trailingSlash,
 	buildFormat,
 	base,
-}: FindRouteToRewrite): [RouteData, URL] {
-	let finalUrl: URL | undefined = undefined;
+}: FindRouteToRewrite): FindRouteToRewriteResult {
+	let newUrl: URL | undefined = undefined;
 	if (payload instanceof URL) {
-		finalUrl = payload;
+		newUrl = payload;
 	} else if (payload instanceof Request) {
-		finalUrl = new URL(payload.url);
+		newUrl = new URL(payload.url);
 	} else {
-		finalUrl = new URL(payload, new URL(request.url).origin);
+		newUrl = new URL(payload, new URL(request.url).origin);
+	}
+	let pathname = newUrl.pathname;
+	if (base !== '/' && newUrl.pathname.startsWith(base)) {
+		pathname = shouldAppendForwardSlash(trailingSlash, buildFormat)
+			? appendForwardSlash(newUrl.pathname)
+			: base !== '/'
+				? removeTrailingForwardSlash(newUrl.pathname)
+				: newUrl.pathname;
+		if (base !== '/') {
+			pathname = pathname.slice(base.length);
+		}
 	}
```

Comment:
Since the outer block already checks `base !== '/'`, could we remove the inner ones?